sources/events.go

```diff
@@ -0,0 +1,131 @@
+// Copyright 2015 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sources
+
+import (
+	"errors"
+
+	kubeapi "github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	kubeclient "github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	kubelabels "github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	kubewatch "github.com/GoogleCloudPlatform/kubernetes/pkg/watch"
+	"github.com/golang/glog"
+)
+
+// EventsSource objects are used to interact with the event source
+// which aggregates all events from a specified Kubernetes API server
+type EventsSource interface {
+	// Returns all new events since GetEvents was last called.
+	GetEvents() ([]kubeapi.Event, error)
+}
+
+// NewEventsSource initializes a new events source and starts a
+// goroutine to watch/fetch event updates.
+func NewEventsSource(client *kubeclient.Client) EventsSource {
+	eventsChan := make(chan eventsUpdate)
+	glog.V(4).Infof("Starting event source")
+	go watchLoop(client.Events(kubeapi.NamespaceAll), eventsChan)
+	glog.V(4).Infof("Finished starting event source")
+	return &eventsSourceImpl{client, eventsChan}
+}
+
+// eventsUpdate is the wrapper object used to pass new events around
+type eventsUpdate struct {
+	events *kubeapi.EventList
+}
+
+// eventsSourceImpl is an implmentation of eventsSource
+type eventsSourceImpl struct {
+	*kubeclient.Client
+	eventsChannel <-chan eventsUpdate
+}
+
+// GetEvents returns all new events since GetEvents was last called.
+func (eventSource *eventsSourceImpl) GetEvents() ([]kubeapi.Event, error) {
+	events := []kubeapi.Event{}
+UpdateLoop:
+	for {
+		select {
+		case eventsUpdate := <-eventSource.eventsChannel:
+			if eventsUpdate.events == nil {
+				return nil, errors.New("Error: recieved a nil event list.")
+			}
+			if eventsUpdate.events.Items == nil {
+				return nil, errors.New("Error: received an event list with nil Items.")
+			}
+			for _, event := range eventsUpdate.events.Items {
+				glog.V(3).Infof("Received new event: %#v\r\n", event)
+				events = append(events, event)
+			}
+		default:
+			break UpdateLoop
```

Comment:
Instead of having two identical return statements, I prefer having one return statement for the method. But, no big deal either way. If you feel strongly, I can change it.